sources/api/prairiedog/src/main.rs

```diff
@@ -253,8 +284,126 @@ fn load_crash_kernel() -> Result<()> {
+        Some(boot_settings) => {
+            info!("Generating initrd boot config from boot settings");
+            trace!("Boot settings: {:?}", boot_settings);
+            let bootconfig = bootconfig::ser::to_string(&boot_settings)
+                .context(error::SerializeBootConfigSnafu)?;
+            trace!("Serializing boot config string: {}", bootconfig);
+            generate_initrd(&bootconfig.into_bytes())?
+        }
+        None => {
+            // If we don't have any boot settings, write out an initrd with default boot config contents
+            trace!("Serializing boot config string: {}", DEFAULT_BOOTCONFIG_STR);
+            generate_initrd(&DEFAULT_BOOTCONFIG_STR.to_string().into_bytes())?
+        }
+    };
+    trace!("Writing initrd image file: {:?}", initrd);
+    tokio::fs::write(BOOTCONFIG_INITRD_PATH, &initrd)
+        .await
+        .context(error::WriteInitrdSnafu)?;
+    Ok(())
+}
+
+/// Retrieves boot config related Bottlerocket settings. If they don't exist in the settings model,
+/// we return `None` instead.
+async fn get_boot_config_settings<P>(socket_path: P) -> Result<Option<BootSettings>>
+where
+    P: AsRef<Path>,
+{
+    let uri = "/settings";
+    let settings: serde_json::Value =
+        schnauzer::get_json(socket_path, uri, Some(("prefix", "boot")))
+            .await
+            .context(error::RetrieveSettingsSnafu)?;
+
+    match settings.get("boot") {
+        None => Ok(None),
+        Some(boot_settings_val) => Ok(Some(
+            serde_json::from_value(boot_settings_val.to_owned()).unwrap(),
+        )),
+    }
+}
+
+/// Reads `/proc/bootconfig` and populates the Bottlerocket boot settings based on the existing boot config data
+pub(crate) async fn generate_boot_settings() -> Result<()> {
+    let proc_bootconfig = match tokio::fs::read_to_string(PROC_BOOTCONFIG).await {
+        Ok(s) => s,
+        Err(e) => {
+            // If there's no `/proc/bootconfig`, then the user hasn't provisioned any kernel boot configuration.
+            if e.kind() == io::ErrorKind::NotFound {
+                // No work to be done
+                return Ok(());
+            }
+            return Err(e).context(error::ReadFileSnafu {
+                path: PROC_BOOTCONFIG,
+            });
+        }
+    };
+    debug!(
+        "Generating kernel boot config settings from `{}`",
+        PROC_BOOTCONFIG
+    );
+    println!("{}", boot_config_to_boot_settings_json(&proc_bootconfig)?);
+
+    Ok(())
+}
+
+/// Given a boot config string, deserialize it to `model::BootSettings` and then serialize it back
+/// out as a JSON string for sundog consumption
+fn boot_config_to_boot_settings_json(bootconfig_str: &str) -> Result<String> {
+    // We'll only send the setting if the existing boot config file fits our settings model
+    let boot_settings =
+        bootconfig::de::from_str::<BootSettings>(bootconfig_str, Some(&["kernel", "init"]))
```

Comment:
Same re-export suggestion here -
```suggestion
        bootconfig::from_str::<BootSettings>(bootconfig_str, Some(&["kernel", "init"]))
```

(Although it doesn't seem idiomatic to take more than just a `str` as input to a `from_str` function.)

Reply:
The more I think about it, the more unnatural it feels to deserialize directly into the `BootSettings` structure like this. That ends up polluting our deserializer with knowledge about the destination shape, which breaks the abstraction this interface is intended to provide.

The trouble is that the maps in the bootconfig file might reach an arbitrary depth; we could end up with:
* `Map<String, String>`
* `Map<Map<String, String>`
* `Map<Map<Map<String, String>`

And so on.

If we had a `Value` type then we could make `BootSettings` handle this, [maybe](https://gist.github.com/bcressey/496abca53539510f5f073f5d8182424e).

But you didn't want a `Value` type and I agree that's pretty complex. I'm just not sure if there's a middle ground. If we can't round-trip somewhat arbitrary data structures to and from the bootconfig format then I'm not confident `serde` is the right interface.

To take a step back - what if we had just two functions and no serde?

First function could turn a `String` read from `/proc/bootconfig` into a `BootSettings` struct, checking each line for either an `init` or `kernel` prefix, stripping that off, splitting the rest of the line on the first " = ", and putting the entry into the corresponding `HashMap<BootConfigKey, Vec<BootConfigValue>>`. We'd use this during settings generation.

Second function could turn a `BootSettings' struct into a `String` in the obvious way. We'd use this during initrd generation.

Do we need more flexibility than that?

Reply:
I agree the interface here isn't the best.

I think that's reasonable. The original thinking was that we'd implement this with serde with the goal to eventually expand it so it can be used elsewhere, e.g. an CLI tool for generating/validating bootconfig files. Maybe that's a case of YAGNI and we can avoid all that complexity and code for now.

I'll go ahead with direct parsing so this PR should slim down a lot more.

Reply:
No worries! I also started out feeling like serde made the most sense. I think the moment it stopped being a natural fit was when we realized that the `BootSettings` struct couldn't just be a pure deserialization because of the arbitrary nesting depth of kernel parameters - but at that point the serde implementation was already done and hard to un-see.

I've certainly learned a lot more about serde from studying this PR, so I consider it a valuable experience and not a mistake. But I agree that YAGNI is the right call.